src/hmem_cuda_gdrcopy.c

```diff
@@ -0,0 +1,380 @@
+	}
+
+	global_gdrcopy_ops.gdr_close = dlsym(gdrapi_handle, "gdr_close");
+	if (!global_gdrcopy_ops.gdr_close) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_close\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	global_gdrcopy_ops.gdr_pin_buffer = dlsym(gdrapi_handle, "gdr_pin_buffer");
+	if (!global_gdrcopy_ops.gdr_pin_buffer) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_pin_buffer\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	global_gdrcopy_ops.gdr_unpin_buffer = dlsym(gdrapi_handle, "gdr_unpin_buffer");
+	if (!global_gdrcopy_ops.gdr_unpin_buffer) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_unpin_buffer\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	global_gdrcopy_ops.gdr_map = dlsym(gdrapi_handle, "gdr_map");
+	if (!global_gdrcopy_ops.gdr_map) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_map\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	global_gdrcopy_ops.gdr_unmap = dlsym(gdrapi_handle, "gdr_unmap");
+	if (!global_gdrcopy_ops.gdr_unmap) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_unmap\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	global_gdrcopy_ops.gdr_copy_to_mapping = dlsym(gdrapi_handle, "gdr_copy_to_mapping");
+	if (!global_gdrcopy_ops.gdr_copy_to_mapping) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_copy_to_mapping\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	global_gdrcopy_ops.gdr_copy_from_mapping = dlsym(gdrapi_handle, "gdr_copy_from_mapping");
+	if (!global_gdrcopy_ops.gdr_copy_from_mapping) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_copy_from_mapping\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	return FI_SUCCESS;
+
+err_dlclose_gdrapi:
+	memset(&global_gdrcopy_ops, 0, sizeof(global_gdrcopy_ops));
+	dlclose(gdrapi_handle);
+	return -FI_ENODATA;
+}
+
+static int cuda_gdrcopy_dl_hmem_cleanup(void)
+{
+	dlclose(gdrapi_handle);
+	return FI_SUCCESS;
+}
+
+#else
+
+static struct gdrcopy_ops global_gdrcopy_ops = {
+	.gdr_open = gdr_open,
+	.gdr_close = gdr_close,
+	.gdr_pin_buffer = gdr_pin_buffer,
+	.gdr_unpin_buffer = gdr_unpin_buffer,
+	.gdr_map = gdr_map,
+	.gdr_unmap = gdr_unmap,
+	.gdr_copy_to_mapping = gdr_copy_to_mapping,
+	.gdr_copy_from_mapping = gdr_copy_from_mapping
+};
+
+static int cuda_gdrcopy_dl_hmem_init(void)
+{
+	return FI_SUCCESS;
+}
+
+static int cuda_gdrcopy_dl_hmem_cleanup(void)
+{
+	return FI_SUCCESS;
+}
+
+#endif /* ENABLE_CUDA_DLOPEN */
+
+int cuda_gdrcopy_hmem_init(void)
+{
+	int err;
+
+	err = cuda_gdrcopy_dl_hmem_init();
+	if (err) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "gdrcopy_dl_hmem_init failed!\n");
+		return -FI_ENOSYS;
+	}
+
+	assert(global_gdrcopy_ops.gdr_open);
+
+	global_gdr = global_gdrcopy_ops.gdr_open();
+	if (!global_gdr) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "gdr_open failed!\n");
+		cuda_gdrcopy_dl_hmem_cleanup();
+		return -FI_ENOMEM;
+	}
+
+	err = pthread_spin_init(&global_gdr_lock, 0);
+	if (err) {
+		assert(global_gdrcopy_ops.gdr_close);
+		global_gdrcopy_ops.gdr_close(global_gdr);
+		cuda_gdrcopy_dl_hmem_cleanup();
+		return err;
```

Comment:
This is unlikely to fail anyway, but the return value should be -err.  For better maintenance, I would jump to a common exit point where cleanup can be unwound.

Reply:
fixed in new revision.